src/sshagent/OpenSSHKey.cpp

```diff
@@ -0,0 +1,481 @@
+    , m_publicData(QList<QByteArray>())
+    , m_privateData(QList<QByteArray>())
+    , m_comment(QString())
+    , m_error(QString())
+{
+
+}
+
+OpenSSHKey::OpenSSHKey(const OpenSSHKey& other)
+    : QObject(nullptr)
+    , m_type(other.m_type)
+    , m_cipherName(other.m_cipherName)
+    , m_kdfName(other.m_kdfName)
+    , m_kdfOptions(other.m_kdfOptions)
+    , m_rawPrivateData(other.m_rawPrivateData)
+    , m_publicData(other.m_publicData)
+    , m_privateData(other.m_privateData)
+    , m_comment(other.m_comment)
+    , m_error(other.m_error)
+{
+
+}
+
+bool OpenSSHKey::operator==(const OpenSSHKey& other) const
+{
+    // close enough for now
+    return (fingerprint() == other.fingerprint());
+}
+
+const QString OpenSSHKey::cipherName() const
+{
+    return m_cipherName;
+}
+
+const QString OpenSSHKey::type() const
+{
+    return m_type;
+}
+
+int OpenSSHKey::keyLength() const
+{
+    if (m_type == "ssh-dss" && m_publicData.length() == 4) {
+        return (m_publicData[0].length() - 1) * 8;
+    } else if (m_type == "ssh-rsa" && m_publicData.length() == 2) {
+        return (m_publicData[1].length() - 1) * 8;
+    } else if (m_type.startsWith("ecdsa-sha2-") && m_publicData.length() == 2) {
+        return (m_publicData[1].length() - 1) * 4;
+    } else if (m_type == "ssh-ed25519" && m_publicData.length() == 1) {
+        return m_publicData[0].length() * 8;
+    }
+
+    return 0;
+}
+
+const QString OpenSSHKey::fingerprint() const
+{
+    QByteArray publicKey;
+    BinaryStream stream(&publicKey);
+
+    stream.writeString(m_type);
+
+    for (QByteArray ba : m_publicData) {
+        stream.writeString(ba);
+    }
+
+    QByteArray rawHash = QCryptographicHash::hash(publicKey, QCryptographicHash::Sha256);
+
+    return "SHA256:" + QString::fromLatin1(rawHash.toBase64(QByteArray::OmitTrailingEquals));
+}
+
+const QString OpenSSHKey::comment() const
+{
+    return m_comment;
+}
+
+const QString OpenSSHKey::publicKey() const
+{
+    QByteArray publicKey;
+    BinaryStream stream(&publicKey);
+
+    stream.writeString(m_type);
+
+    for (QByteArray ba : m_publicData) {
+        stream.writeString(ba);
+    }
+
+    return m_type + " " + QString::fromLatin1(publicKey.toBase64()) + " " + m_comment;
+}
+
+const QString OpenSSHKey::errorString() const
+{
+    return m_error;
+}
+
+void OpenSSHKey::setType(const QString& type)
+{
+    m_type = type;
+}
+
+void OpenSSHKey::setPublicData(const QList<QByteArray>& data)
+{
+    m_publicData = data;
+}
+
+void OpenSSHKey::setPrivateData(const QList<QByteArray>& data)
+{
+    m_privateData = data;
+}
+
+void OpenSSHKey::setComment(const QString& comment)
+{
+    m_comment = comment;
+}
+
+void OpenSSHKey::clearPrivate()
+{
+    m_rawPrivateData.clear();
+    m_privateData.clear();
+}
+
+bool OpenSSHKey::parsePEM(const QByteArray& in, QByteArray& out)
+{
+    QString pem = QString::fromLatin1(in);
+    QStringList rows = pem.split(QRegularExpression("[\r\n]"), QString::SkipEmptyParts);
```

Comment:
This regular expression should be `"\r?\n"`, otherwise it will split twice on CRLF.

Reply:
It doesn't matter with `QString::SkipEmptyParts` and it also handles plain `\r` line endings.

Reply:
Which would be `"(?:\r|\r?\n)"`.
I like precised regular expressions. ;-)

Reply:
I will bet my unborn child that `[\r\n]` is faster even with the join! 🔥

Reply:
It's negligibly faster, yes. For LF and CRLF texts, `(?:\r?\n|\r)` is also slightly faster than `(?:\r|\r?\n)`, because it doesn't require as much backtracking. `[\r\n]` is constant in all cases and scales with the number of empty parts it needs to skip.
I won't start a war over this, do as you like. 😉

Reply:
Actually, I just benchmarked it. I ran 1,000,000 iterations with three splits. It's exactly as I said: `[\r\n]` is the fastest for a text with only CR and LF line endings (5.3 seconds).
`(?:\r?\n|\r)` on a pure or mixed CR/CRLF /LF text is about 6.4 seconds. Worst case is `(?:\r|\r?\n)` on a pure CRLF or LF text with 8.7 seconds.
On a pure CRLF text, `(?:\r?\n|\r)` is the fastest with 6.4 seconds vs. `[\r\n]` with 7.4 seconds.

Reply:
I love it that you guys benchmarked this, but in most use cases, we're looking at milliseconds difference here - no matter which method used. Full marks for impractical applications though ;)

Reply:
You won't find a measurable difference if you don't run at least 100k or 1m iterations. But I can't lose a bet like this. 😉

Reply:
🏳️